src/main/scala/analysis/NonReturningFunctions.scala

```diff
@@ -0,0 +1,139 @@
+import scala.collection.mutable
+import scala.collection.mutable.ArrayBuffer
+import scala.collection.parallel.CollectionConverters.*
+
+class NonReturningFunctions {
+
+  def transform(procedures: ArrayBuffer[Procedure]): Unit = {
+    val blocksToRemove: Queue[String] = Queue()
+    val mapJumpsToBlocks: Map[String, ArrayBuffer[(Jump, Block)]] = Map()
+    val mapBlocksToProcedure: Map[String, (Procedure, Integer)] = Map()
+
+
+    // Check if the goto is part of a endless loop by checking to see if the index its jumping to is earlier on in the method,
+    // and that any jumps between the index its jumping to, and the index of the instruction, don't jump outside of that range.
+    // This method checks for both continue (which jumps back to the while loop, and break (which jumps outside of loops).
+    def isEndlessLoop(proc: Procedure, goTo: GoTo, index: Integer): Boolean = {
+
+      if (goTo.condition.isEmpty && mapBlocksToProcedure.contains(goTo.target.label) && mapBlocksToProcedure(goTo.target.label)._2 < index) {
+
+        val (_, idx) = mapBlocksToProcedure(goTo.target.label)
+
+        for (loopIndex <- idx.toInt to index.toInt) {
+          val blockAtLoopIndex = proc.blocks(loopIndex)
+          for (loopJump <- blockAtLoopIndex.jumps) {
+            loopJump match {
+              case loopGoTo: GoTo =>
+                if (!mapBlocksToProcedure.contains(loopGoTo.target.label) || mapBlocksToProcedure(loopGoTo.target.label)._2 >= index)
+                  return false
+              case call: IndirectCall =>
+                if (call.target.name == "R30") {
+                  return false
+                }
+              case _ =>
+            }
+          }
+
+        }
+        return true
+      }
+      false
+    }
+
+    // look into each procedure, and calculate the number of return statements in each block
+    // and create maps between jumps and blocks, and blocks and procedures.
+    // this also looks at endless loops, and removes unreachable code after endless blocks
+    for (proc <- procedures) {
+
+      for ((block, index) <- proc.blocks.zipWithIndex) {
+
+        mapBlocksToProcedure.addOne(block.label, (proc, index))
+        for (jump <- block.jumps) {
+
+          jump match {
+            case call: IndirectCall =>
+              if (call.target.name == "R30") {
+                block.countOfReturnStatements += 1
+              }
+            case directCall: DirectCall =>
+              mapJumpsToBlocks.put(directCall.target.name, mapJumpsToBlocks.getOrElse(directCall.target.name, ArrayBuffer()).addOne((directCall, block)))
+            case goTo: GoTo =>
+              mapJumpsToBlocks.put(goTo.target.label, mapJumpsToBlocks.getOrElse(goTo.target.label, ArrayBuffer()).addOne((goTo, block)))
+              if (proc.blocks.length > index && isEndlessLoop(proc, goTo, index)) {
+                blocksToRemove.enqueue(proc.blocks(index + 1).label)

```

Comment:
This line throws an IndexOutOfBoundsException for the example complex.

Reply:
Fixed